homeassistant/components/switch/smartthings.py

```diff
@@ -0,0 +1,44 @@
+"""
+Support for switches through the SmartThings cloud API.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/switch.smartthings/
+"""
+from homeassistant.components.smartthings import SmartThingsEntity
+from homeassistant.components.smartthings.const import (
+    DATA_BROKERS, DOMAIN as SMARTTHINGS_DOMAIN)
+from homeassistant.components.switch import ToggleEntity
+
+DEPENDENCIES = ['smartthings']
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Platform uses config entry setup."""
+    pass
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add switches for a config entry."""
+    broker = hass.data[SMARTTHINGS_DOMAIN][DATA_BROKERS][config_entry.entry_id]
+    async_add_entities(
+        [SmartThingsSwitch(device) for device in broker.switches])
+
+
+class SmartThingsSwitch(SmartThingsEntity, ToggleEntity):
+    """Define a SmartThings switch."""
+
+    async def async_turn_off(self, **kwargs) -> None:
+        """Turn the switch off."""
+        await self._device.switch_off(True)
+        self.async_schedule_update_ha_state()
```

Comment:
Is this necessary? Wouldn't SmartThings push an update to us? And if we call device.switch_off, do they change the state optimistically internally?

Reply:
It is pushed, but only after the device has _completed_ transition to the state.  This results in a bad UI experience if the device doesn't instantly turn off, such as a z-wave dimmer.  In the front-end, the toggle flips back to `on` until the push comes in.  Passing `True` to `switch_off` updates the state optimistically, in anticipation of the push update.

Reply:
Things like that should not be solved in this layer, but solved at the frontend level (where we do have a 2s optimistic toggle)

Reply:
Does the UI layer provide a mechanism to solve this? (i.e. for these entities, use an optimistic state when actuated?).  A static 2s won't work in dimmer cases.  For example the default transition on GE ZWave Dimmers is ~ 2.4 seconds -- and is configurable to be much longer.  Also using a duration parameter with turn_on (in a light platform) > 2s...

Reply:
Can we leave this in for now until we can figure something better out?  I don't see any harm.  Happy to help explore other options.